packages/automated-dispute/src/types/eventsV2.types.ts

```diff
@@ -0,0 +1,89 @@
+import { UnixTimestamp } from "@ebo-agent/shared";
+import { AbiParametersToPrimitiveTypes, ExtractAbiEvent, ExtractAbiEventNames } from "abitype";
+import { Log } from "viem";
+
+import { eboAbi } from "../abis/ebo.js";
+import { ProposalId, ProposalStatus } from "./index.js";
+
+export type EboEventNameV2 = Extract<
+    ExtractAbiEventNames<typeof eboAbi>,
+    | "ProposalCreated"
+    | "ProposalDisputed"
+    | "ProposalFinalized"
+    | "PledgedForDispute"
+    | "PledgedAgainstDispute"
+    | "DisputeEscalated"
+    | "DisputeResolved"
+>;
+
+type AbiEventToType<T extends readonly unknown[]> = {
+    [K in T[number] as K extends { name: infer Name extends string }
+        ? Name extends `_${infer CleanName}`
+            ? CleanName
+            : Name
+        : never]: K extends {
+        type: string;
+    }
+        ? AbiParametersToPrimitiveTypes<[K]>[0]
+        : never;
+};
+
+export type ProposalCreatedV2 = AbiEventToType<
+    ExtractAbiEvent<typeof eboAbi, "ProposalCreated">["inputs"]
+>;
+
+export type ProposalDisputedV2 = AbiEventToType<
+    ExtractAbiEvent<typeof eboAbi, "ProposalDisputed">["inputs"]
+>;
+
+export type ProposalFinalizedV2 = AbiEventToType<
+    ExtractAbiEvent<typeof eboAbi, "ProposalFinalized">["inputs"]
+>;
+
+export type PledgedForDisputeV2 = AbiEventToType<
+    ExtractAbiEvent<typeof eboAbi, "PledgedForDispute">["inputs"]
+>;
+
+export type PledgedAgainstDisputeV2 = AbiEventToType<
+    ExtractAbiEvent<typeof eboAbi, "PledgedAgainstDispute">["inputs"]
+>;
+
+export type DisputeEscalatedV2 = AbiEventToType<
+    ExtractAbiEvent<typeof eboAbi, "DisputeEscalated">["inputs"]
+>;
+
+export type DisputeResolvedV2 = Omit<
+    AbiEventToType<ExtractAbiEvent<typeof eboAbi, "DisputeResolved">["inputs"]>,
+    "resolutionStatus"
+> & {
+    resolutionStatus: ProposalStatus;
+};
+
+export type EboEventDataV2<T extends EboEventNameV2> = T extends "ProposalCreated"
+    ? ProposalCreatedV2
+    : T extends "ProposalDisputed"
+      ? ProposalDisputedV2
+      : T extends "ProposalFinalized"
+        ? ProposalFinalizedV2
+        : T extends "PledgedForDispute"
+          ? PledgedForDisputeV2
+          : T extends "PledgedAgainstDispute"
+            ? PledgedAgainstDisputeV2
+            : T extends "DisputeEscalated"
+              ? DisputeEscalatedV2
+              : T extends "DisputeResolved"
+                ? DisputeResolvedV2
+                : never;
+
+export type EboEventV2<T extends EboEventNameV2> = {
+    name: T;
+    /** Block's number this event was logged on */
+    blockNumber: bigint;
+    /** Log index relative to the block this event was logged on */
+    logIndex: number;
+    /** Block's timestamp this event was logged on */
+    timestamp: UnixTimestamp;
+    rawLog?: Log;
+    proposalId: ProposalId; // Field to use to route events to actors
+    metadata: EboEventDataV2<T>;
```

Comment:
shouldn't be like this ? 

```suggestion
    params: EboEventDataV2<T>;
```

Reply:
It'd be great to rename `metadata`